libraries/botframework-config/src/botConfiguration.ts

```diff
@@ -0,0 +1,230 @@
+/**
+ * Copyright(c) Microsoft Corporation.All rights reserved.
+ * Licensed under the MIT License.
+ */
+import * as crypto from 'crypto';
+import * as fsx from 'fs-extra';
+import { Enumerable, List } from 'linq-collections';
+import * as path from 'path';
+import * as process from 'process';
+import * as txtfile from 'read-text-file';
+import * as uuid from 'uuid';
+import { decryptString, encryptString } from './encrypt';
+import { AzureBotService, ConnectedService, DispatchService, EndpointService, FileService, LuisService, QnaMakerService } from './models';
+import { IAzureBotService, IBotConfiguration, IConnectedService, IDispatchService, IEndpointService, IFileService, ILuisService, IQnAService, ServiceTypes } from './schema';
+
+
+interface internalBotConfig {
+    location?: string;
+}
+
+export class BotConfiguration implements Partial<IBotConfiguration> {
+    // internal is not serialized
+    private internal: internalBotConfig = {
+    };
+
+    public name: string = '';
+    public description: string = '';
+    public services: IConnectedService[] = [];
+    public secretKey = '';
+
+    constructor() {
+    }
+
+    public static fromJSON(source: Partial<IBotConfiguration> = {}): BotConfiguration {
+        let { name = '', description = '', secretKey = '', services = [] } = source;
+        services = services.slice().map(BotConfiguration.serviceFromJSON);
+        const botConfig = new BotConfiguration();
+        Object.assign(botConfig, { services, description, name, secretKey });
+        return botConfig;
+    }
+
+    public toJSON(): Partial<IBotConfiguration> {
+        const { name, description, services, secretKey } = this;
+        return { name, description, services, secretKey };
+    }
+
+    public static async loadBotFromFolder(folder?: string, secret?: string): Promise<BotConfiguration> {
+        let files = Enumerable.fromSource(await fsx.readdir(folder || process.cwd()))
```

Comment:
Consider using [node-klaw](https://github.com/jprichardson/node-klaw) for directory contents traversal. It's very lightweight, trusted and widely known in the js community.